src/rust/mqtt-publisher/src/main.rs

```diff
@@ -0,0 +1,230 @@
+                    .publish(
+                        format!(
+                            "place_market_order/{}/{}/{}/{}",
+                            data.market_id, data.user, data.custodian_id, data.integrator
+                        ),
+                        QoS::AtLeastOnce,
+                        false,
+                        serde_json::to_string(&data)?,
+                    )
+                    .await?;
+            }
+            "place_swap_order" => {
+                let data: PlaceSwapOrderNotif = serde_json::from_str(notification.payload())?;
+                client
+                    .publish(
+                        format!(
+                            "place_swap_order/{}/{}/{}",
+                            data.market_id, data.integrator, data.signing_account
+                        ),
+                        QoS::AtLeastOnce,
+                        false,
+                        serde_json::to_string(&data)?,
+                    )
+                    .await?;
+            }
+            "change_order_size" => {
+                let data: ChangeOrderSizeNotif = serde_json::from_str(notification.payload())?;
+                client
+                    .publish(
+                        format!(
+                            "change_order_size/{}/{}/{}",
+                            data.market_id, data.user, data.custodian_id
+                        ),
+                        QoS::AtLeastOnce,
+                        false,
+                        serde_json::to_string(&data)?,
+                    )
+                    .await?;
+            }
+            "cancel_order" => {
+                let data: CancelOrderNotif = serde_json::from_str(notification.payload())?;
+                client
+                    .publish(
+                        format!(
+                            "cancel_order_notif/{}/{}/{}",
+                            data.market_id, data.user, data.custodian_id
+                        ),
+                        QoS::AtLeastOnce,
+                        false,
+                        serde_json::to_string(&data)?,
+                    )
+                    .await?;
+            }
+            "fill" => {
+                let data: FillNotif = serde_json::from_str(notification.payload())?;
+                client
+                    .publish(
+                        format!(
+                            "fill_notif/{}/{}/{}",
+                            data.market_id, data.maker_address, data.maker_custodian_id
+                        ),
+                        QoS::AtLeastOnce,
+                        false,
+                        serde_json::to_string(&data)?,
+                    )
+                    .await?;
+                client
+                    .publish(
+                        format!(
+                            "fill_notif/{}/{}/{}",
+                            data.market_id, data.taker_address, data.taker_custodian_id
```

Comment:
I understand that we have duplicity in fill events, but can you explain why there are two "clients" here?

Is it so that you can subscribe to your fill events as a maker on one channel, and as a taker on another channel?

Reply:
It is so that when you subscribe to fills for one address, you get fills no matter if you're the maker or taker.